pkg/operator/certrotationcontroller/satokensigner_controller.go

```diff
@@ -270,6 +281,29 @@ func (c *SATokenSignerController) syncWorker() error {
 	return nil
 }
 
+//isValidKeyPair checks if a public/private keys match i.e if the tokens signed
+//with private key can be validated using public key. if "pubKeyData" or "privKeyData"
+//contains multiple keys, it tries to validate the first available keys in byte slices
+func isValidKeyPair(pubKeyData, privKeyData []byte) (bool, error) {
+	privKey, err := keyutil.ParsePrivateKeyPEM(privKeyData)
+	if err != nil {
+		return false, err
+	}
+	rsaPrivateKey, ok := privKey.(*rsa.PrivateKey)
+	if !ok {
+		return false, fmt.Errorf("private key is not of rsa type")
+	}
+	pubKeys, err := keyutil.ParsePublicKeysPEM(pubKeyData)
+	if err != nil {
+		return false, err
+	}
+	wantRSAPublicKey, ok := pubKeys[0].(*rsa.PublicKey)
+	if !ok {
+		return false, fmt.Errorf("public key is not if rsa type")
+	}
+	return reflect.DeepEqual(rsaPrivateKey.PublicKey, *wantRSAPublicKey), nil
```

Comment:
this is the raw rsa key blob?

Reply:
It is the `PublicKey` struct https://golang.org/pkg/crypto/rsa/#PublicKey

Reply:
Maybe add a comment reminding future maintainers that the private key embeds public key and that embedded key must match the provided public key?

Reply:
golang private key type des embed it. I am not sure that expectation is placed on `saTokenSigner.Data["service-account.pub"]`. Generally anything that is named `.pub` contains the public key only and is not considered to be secret. Combined private+secret keys are named `.pem` and then you wouldn't need `saTokenSigner.Data["service-account.key"]`, right? I didn't go back tracing this in the code but it looks suspicious but maybe you can trace it and prove me wrong.

Reply:
There is no reason for the signing secret to contain the public key on its own. It can always be derived from the private key, which embeds it. I was initially ignorant of this fact, and that appears to be the case for the original implementers as well or they wouldn't have bothered writing the public key to the secret since it can always be sourced from the private key before being published (e.g. to a config map for the apiserver to consume).